File: src/android/TalkingDataAppCpaPlugin.java
package com.talkingdata.adtracking;

import android.content.Context;
import android.text.TextUtils;
import com.tendcloud.appcpa.TDSearch;

import com.tendcloud.appcpa.Order;
import com.tendcloud.appcpa.ShoppingCart;
import com.tendcloud.appcpa.TDTransaction;
import com.tendcloud.appcpa.TalkingDataAppCpa;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;


public class TalkingDataAppCpaPlugin extends CordovaPlugin {
    private Context ctx;

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
        this.ctx = cordova.getActivity().getApplicationContext();
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if ("setVerboseLogDisable".equals(action)) {
            TalkingDataAppCpa.setVerboseLogDisable();
            return true;
        } else if ("getDeviceId".equals(action)) {
            String deviceId = TalkingDataAppCpa.getDeviceId(ctx);
            callbackContext.success(deviceId);
            return true;
        } else if ("getOAID".equals(action)) {
            String oaid = TalkingDataAppCpa.getOAID(ctx);
            callbackContext.success(oaid);
            return true;
        } else if ("onRegister".equals(action)) {
            String profileId = args.getString(0);
            String invitationCode = args.getString(1);
            TalkingDataAppCpa.onRegister(profileId, invitationCode);
            return true;
        } else if ("onLogin".equals(action)) {
            String profileId = args.getString(0);
            TalkingDataAppCpa.onLogin(profileId);
            return true;
        } else if (action.startsWith("onCreateCard")) {
            String profileId = args.getString(0);
            String method = args.getString(1);
            String content = args.getString(2);
            TalkingDataAppCpa.onCreateCard(profileId, method, content);
            return true;
        } else if ("onReceiveDeepLink".equals(action)) {
            String link = args.getString(0);
            TalkingDataAppCpa.onReceiveDeepLink(link);
            return true;
        } else if (action.startsWith("onFavorite")) {
            String category = args.getString(0);
            String content = args.getString(1);
            TalkingDataAppCpa.onFavorite(category, content);
            return true;
        } else if (action.startsWith("onShare")) {
            String profileId = args.getString(0);
            String content = args.getString(1);
            TalkingDataAppCpa.onShare(profileId, content);
            return true;
        } else if (action.startsWith("onPunch")) {
            String profileId = args.getString(0);
            String punchId  = args.getString(1);
            TalkingDataAppCpa.onPunch(profileId, punchId);
            return true;
        } else if (action.startsWith("onSearch")) {
            String search = args.getString(0);
            TalkingDataAppCpa.onSearch(stringToSearch(search));
            return true;
        } else if (action.startsWith("onContact")) {
            String profileId = args.getString(0);
            String content = args.getString(1);
            TalkingDataAppCpa.onContact(profileId, content);
            return true;
        } else if ("onPay".equals(action)) {
            String profileId = args.getString(0);
            String orderId = args.getString(1);
            int amount = args.getInt(2);
            String currencyType = args.getString(3);
            String payType = args.getString(4);
            TalkingDataAppCpa.onPay(profileId, orderId, amount, currencyType, payType);
            return true;
        } else if (action.startsWith("onChargeBack")) {
            String profileId = args.getString(0);
            String orderId = args.getString(1);
            String reason = args.getString(2);
            String type = args.getString(3);
            TalkingDataAppCpa.onChargeBack(profileId, orderId, reason, type);
            return true;
        } else if (action.startsWith("onReservation")) {
            String profileId = args.getString(0);
            String reservationId = args.getString(1);
            String category = args.getString(2);
            int amount = args.getInt(3);
            String term = args.getString(4);
            TalkingDataAppCpa.onReservation(profileId, reservationId, category, amount, term);
            return true;
        } else if (action.startsWith("onBooking")) {
            String profileId = args.getString(0);
            String bookingId = args.getString(1);
            String category = args.getString(2);
            int amount = args.getInt(3);
            String content = args.getString(4);
            TalkingDataAppCpa.onBooking(profileId, bookingId, category, amount, content);
            return true;
        } else if ("onViewItem".equals(action)) {
            String itemId = args.getString(0);
            String category = args.getString(1);
            String name = args.getString(2);
            int unitPrice = args.getInt(3);
            TalkingDataAppCpa.onViewItem(itemId, category, name, unitPrice);
            return true;
        } else if ("onAddItemToShoppingCart".equals(action)) {
            String itemId = args.getString(0);
            String category = args.getString(1);
            String name = args.getString(2);
            int unitPrice = args.getInt(3);
            int amount = args.getInt(4);
            TalkingDataAppCpa.onAddItemToShoppingCart(itemId, category, name, unitPrice, amount);
            return true;
        } else if ("onViewShoppingCart".equals(action)) {
            String shoppingCartStr = args.getString(0);
            ShoppingCart shoppingCart = this.stringToShoppingCart(shoppingCartStr);
            TalkingDataAppCpa.onViewShoppingCart(shoppingCart);
            return true;
        } else if ("onPlaceOrder".equals(action)) {
            String profileId = args.getString(0);
            String orderStr = args.getString(1);
            Order order = this.stringToOrder(orderStr);
            TalkingDataAppCpa.onPlaceOrder(profileId, order);
            return true;
        } else if ("onOrderPaySucc".equals(action)) {
            String profileId = args.getString(0);
            String orderId = args.getString(1);
            int amount = args.getInt(2);
            String currencyType = args.getString(3);
            String payType = args.getString(4);
            TalkingDataAppCpa.onOrderPaySucc(profileId, orderId, amount, currencyType, payType);
            return true;
        } else if (action.startsWith("onCredit")) {
            String profileId = args.getString(0);
            int amount = args.getInt(1);
            String content = args.getString(2);
            TalkingDataAppCpa.onCredit(profileId, amount, content);
            return true;
        } else if (action.startsWith("onTransaction")) {
            String profileId = args.getString(0);
            String transaction = args.getString(1);
            TalkingDataAppCpa.onTransaction(profileId, stringToTransaction(transaction));
            return true;
        } else if ("onCreateRole".equals(action)) {
            String roleName = args.getString(0);
            TalkingDataAppCpa.onCreateRole(roleName);
            return true;
        } else if (action.startsWith("onLevelPass")) {
            String profileId = args.getString(0);
            String levelId = args.getString(1);
            TalkingDataAppCpa.onLevelPass(profileId, levelId);
            return true;
        } else if (action.startsWith("onGuideFinished")) {
            String profileId = args.getString(0);
            String content = args.getString(1);
            TalkingDataAppCpa.onGuideFinished(profileId, content);
            return true;
        } else if (action.startsWith("onLearn")) {
            String profileId = args.getString(0);
            String course = args.getString(1);
            long begin = args.getLong(2);
            int duration = args.getInt(3);
            TalkingDataAppCpa.onLearn(profileId, course, begin, duration);
            return true;
        } else if (action.startsWith("onPreviewFinished")) {
            String profileId = args.getString(0);
            String content = args.getString(1);
            TalkingDataAppCpa.onPreviewFinished(profileId, content);
            return true;
        } else if (action.startsWith("onRead")) {
            String profileId = args.getString(0);
            String book = args.getString(1);
            long begin = args.getLong(2);
            int duration = args.getInt(3);
            TalkingDataAppCpa.onRead(profileId, book, begin, duration);
            return true;
        } else if (action.startsWith("onFreeFinished")) {
            String profileId = args.getString(0);
            String content = args.getString(1);
            TalkingDataAppCpa.onFreeFinished(profileId, content);
            return true;
        } else if (action.startsWith("onAchievementUnlock")) {
            String profileId = args.getString(0);
            String achievementId = args.getString(1);
            TalkingDataAppCpa.onAchievementUnlock(profileId, achievementId);
            return true;
        } else if (action.startsWith("onBrowse")) {
            String profileId = args.getString(0);
            String content = args.getString(1);
            long begin = args.getLong(2);
            int duration = args.getInt(3);
            TalkingDataAppCpa.onBrowse(profileId, content, begin, duration);
            return true;
        } else if (action.startsWith("onTrialFinished")) {
            String profileId = args.getString(0);
            String content = args.getString(1);
            TalkingDataAppCpa.onTrialFinished(profileId, content);
            return true;
        } else if (action.startsWith("onCustEvent")) {
            try {
                Method onCustEvent = TalkingDataAppCpa.class.getDeclaredMethod(action);
                onCustEvent.invoke(null);
                return true;
            } catch (NoSuchMethodException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            } catch (InvocationTargetException e) {
                e.printStackTrace();
            }
        }
        return false;
    }

    private TDSearch stringToSearch(String json) {
        try {
            JSONObject searchJson = new JSONObject(json);
            TDSearch search = TDSearch.createAdSearch();
            search.setCategory(searchJson.optString("category", null));
            search.setContent(searchJson.optString("content", null));
            search.setItemId(searchJson.optString("itemId", null));
            search.setItemLocationId(searchJson.optString("itemLocationId", null));
            search.setDestination(searchJson.optString("destination", null));
            search.setOrigin(searchJson.optString("origin", null));
            if (searchJson.has("startDate")) {
                search.setStartDate(searchJson.optLong("startDate", 0));
            }
            if (searchJson.has("endDate")) {
                search.setEndDate(searchJson.optLong("endDate", 0));
            }
            return search;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    private Order stringToOrder(String orderStr) {
        try {
            JSONObject orderJson = new JSONObject(orderStr);
            Order order = Order.createOrder(orderJson.getString("orderId"), orderJson.getInt("total"), orderJson.getString("currencyType"));
            JSONArray items = orderJson.getJSONArray("items");
            for (int i = 0; i < items.length(); i++) {
                JSONObject item = items.getJSONObject(i);
                order.addItem(item.getString("itemId"), item.getString("category"), item.getString("name"), item.getInt("unitPrice"), item.getInt("amount"));
            }
            return order;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    private ShoppingCart stringToShoppingCart(String shoppingCartStr) {
        try {
            JSONObject shoppingCartJson = new JSONObject(shoppingCartStr);
            ShoppingCart shoppingCart = ShoppingCart.createShoppingCart();
            JSONArray items = shoppingCartJson.getJSONArray("items");
            for (int i = 0; i < items.length(); i++) {
                JSONObject item = items.getJSONObject(i);
                shoppingCart.addItem(item.getString("itemId"), item.getString("category"), item.getString("name"), item.getInt("unitPrice"), item.getInt("amount"));
            }
            return shoppingCart;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    private TDTransaction stringToTransaction(String json) {
        try {
            JSONObject transactionJson = new JSONObject(json);
            TDTransaction transaction = TDTransaction.createTDTransaction();
            transaction.setTransactionId(transactionJson.optString("transactionId", null));
            transaction.setCategory(transactionJson.optString("category", null));
            transaction.setAmount(transactionJson.optInt("amount", 0));
            transaction.setPersonA(transactionJson.optString("personA", null));
            transaction.setPersonB(transactionJson.optString("personB", null));
            transaction.setStartDate(transactionJson.optLong("startDate", 0));
            transaction.setEndDate(transactionJson.optLong("endDate", 0));
            transaction.setContent(transactionJson.optString("content", null));
            transaction.setCurrencyType(transactionJson.optString("currencyType", null));
            return transaction;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: src/ios/TalkingDataAppCpaPlugin.m
#import "TalkingDataAppCpaPlugin.h"
#import "TalkingDataAppCpa.h"


@implementation TalkingDataAppCpaPlugin

- (void)setVerboseLogDisable:(CDVInvokedUrlCommand*)command {
    [TalkingDataAppCpa setVerboseLogDisable];
}

- (void)getDeviceId:(CDVInvokedUrlCommand*)command {
    NSString *deviceId = [TalkingDataAppCpa getDeviceId];
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:deviceId];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)onRegister:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *invivateCode = [command.arguments objectAtIndex:1];
    if (![invivateCode isKindOfClass:[NSString class]]) {
        invivateCode = nil;
    }
    
    [TalkingDataAppCpa onRegister:profileId invitationCode:invivateCode];
}

- (void)onLogin:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    [TalkingDataAppCpa onLogin:profileId];
}

- (void)onCreateCard:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *method = [command.arguments objectAtIndex:1];
    if (![method isKindOfClass:[NSString class]]) {
        method = nil;
    }
    
    NSString *content = [command.arguments objectAtIndex:2];
    if (![content isKindOfClass:[NSString class]]) {
        content = nil;
    }
    
    [TalkingDataAppCpa onCreateCard:profileId method:method content:content];
}

- (void)onReceiveDeepLink:(CDVInvokedUrlCommand*)command {
    NSString *arg0 = [command.arguments objectAtIndex:0];
    if (![arg0 isKindOfClass:[NSString class]]) {
        return;
    }
    NSURL *url = [NSURL URLWithString:arg0];
    
    [TalkingDataAppCpa onReceiveDeepLink:url];
}

- (void)onFavorite:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *content = [command.arguments objectAtIndex:1];
    if (![content isKindOfClass:[NSString class]]) {
        content = nil;
    }
    
    [TalkingDataAppCpa onFavorite:profileId content:content];
}

- (void)onShare:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *content = [command.arguments objectAtIndex:1];
    if (![content isKindOfClass:[NSString class]]) {
        content = nil;
    }
    
    [TalkingDataAppCpa onShare:profileId content:content];
}

- (void)onPunch:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *punchId = [command.arguments objectAtIndex:1];
    if (![punchId isKindOfClass:[NSString class]]) {
        punchId = nil;
    }
    
    [TalkingDataAppCpa onPunch:profileId punchId:punchId];
}

- (void)onSearch:(CDVInvokedUrlCommand*)command {
    NSString *arg0 = [command.arguments objectAtIndex:0];
    TDSearch *search = [self stringToSearch:arg0];
    
    [TalkingDataAppCpa onSearch:search];
}

- (void)onContact:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *content = [command.arguments objectAtIndex:1];
    if (![content isKindOfClass:[NSString class]]) {
        content = nil;
    }
    
    [TalkingDataAppCpa onContact:profileId content:content];
}

- (void)onPay:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *orderId = [command.arguments objectAtIndex:1];
    if (![orderId isKindOfClass:[NSString class]]) {
        orderId = nil;
    }
    
    int amount = 0;
    NSNumber *arg2 = [command.arguments objectAtIndex:2];
    if ([arg2 isKindOfClass:[NSNumber class]]) {
        amount = [arg2 intValue];
    }
    
    NSString *currencyType = [command.arguments objectAtIndex:3];
    if (![currencyType isKindOfClass:[NSString class]]) {
        currencyType = nil;
    }
    
    NSString *payType = [command.arguments objectAtIndex:4];
    if (![payType isKindOfClass:[NSString class]]) {
        payType = nil;
    }
    
    [TalkingDataAppCpa onPay:profileId withOrderId:orderId withAmount:amount withCurrencyType:currencyType withPayType:payType];
}

- (void)onChargeBack:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *orderId = [command.arguments objectAtIndex:1];
    if (![orderId isKindOfClass:[NSString class]]) {
        orderId = nil;
    }
    
    NSString *reason = [command.arguments objectAtIndex:2];
    if (![reason isKindOfClass:[NSString class]]) {
        reason = nil;
    }
    
    NSString *type = [command.arguments objectAtIndex:3];
    if (![type isKindOfClass:[NSString class]]) {
        type = nil;
    }
    
    [TalkingDataAppCpa onChargeBack:profileId orderId:orderId reason:reason type:type];
}

- (void)onReservation:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *reservationId = [command.arguments objectAtIndex:1];
    if (![reservationId isKindOfClass:[NSString class]]) {
        reservationId = nil;
    }
    
    NSString *category = [command.arguments objectAtIndex:2];
    if (![category isKindOfClass:[NSString class]]) {
        category = nil;
    }
    
    int amount = 0;
    NSNumber *arg3 = [command.arguments objectAtIndex:3];
    if ([arg3 isKindOfClass:[NSNumber class]]) {
        amount = [arg3 intValue];
    }
    
    NSString *term = [command.arguments objectAtIndex:4];
    if (![term isKindOfClass:[NSString class]]) {
        term = nil;
    }
    
    [TalkingDataAppCpa onReservation:profileId reservationId:reservationId category:category amount:amount term:term];
}

- (void)onBooking:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *bookingId = [command.arguments objectAtIndex:1];
    if (![bookingId isKindOfClass:[NSString class]]) {
        bookingId = nil;
    }
    
    NSString *category = [command.arguments objectAtIndex:2];
    if (![category isKindOfClass:[NSString class]]) {
        category = nil;
    }
    
    int amount = 0;
    NSNumber *arg3 = [command.arguments objectAtIndex:3];
    if ([arg3 isKindOfClass:[NSNumber class]]) {
        amount = [arg3 intValue];
    }
    
    NSString *content = [command.arguments objectAtIndex:4];
    if (![content isKindOfClass:[NSString class]]) {
        content = nil;
    }
    
    [TalkingDataAppCpa onBooking:profileId bookingId:bookingId category:category amount:amount content:content];
}

- (void)onViewItem:(CDVInvokedUrlCommand*)command {
    NSString *itemId = [command.arguments objectAtIndex:0];
    if (![itemId isKindOfClass:[NSString class]]) {
        itemId = nil;
    }
    
    NSString *category = [command.arguments objectAtIndex:1];
    if (![category isKindOfClass:[NSString class]]) {
        category = nil;
    }
    
    NSString *name = [command.arguments objectAtIndex:2];
    if (![name isKindOfClass:[NSString class]]) {
        name = nil;
    }
    
    int unitPrice = 0;
    NSNumber *arg3 = [command.arguments objectAtIndex:3];
    if ([arg3 isKindOfClass:[NSNumber class]]) {
        unitPrice = [arg3 intValue];
    }
    
    [TalkingDataAppCpa onViewItemWithCategory:category itemId:itemId name:name unitPrice:unitPrice];
}

- (void)onAddItemToShoppingCart:(CDVInvokedUrlCommand*)command {
    NSString *itemId = [command.arguments objectAtIndex:0];
    if (![itemId isKindOfClass:[NSString class]]) {
        itemId = nil;
    }
    
    NSString *category = [command.arguments objectAtIndex:1];
    if (![category isKindOfClass:[NSString class]]) {
        category = nil;
    }
    
    NSString *name = [command.arguments objectAtIndex:2];
    if (![name isKindOfClass:[NSString class]]) {
        name = nil;
    }
    
    int unitPrice = 0;
    NSNumber *arg3 = [command.arguments objectAtIndex:3];
    if ([arg3 isKindOfClass:[NSNumber class]]) {
        unitPrice = [arg3 intValue];
    }
    
    int amount = 0;
    NSNumber *arg4 = [command.arguments objectAtIndex:4];
    if ([arg4 isKindOfClass:[NSNumber class]]) {
        amount = [arg4 intValue];
    }
    
    [TalkingDataAppCpa onAddItemToShoppingCartWithCategory:category itemId:itemId name:name unitPrice:unitPrice amount:amount];
}

- (void)onViewShoppingCart:(CDVInvokedUrlCommand*)command {
    NSString *arg0 = [command.arguments objectAtIndex:0];
    TDShoppingCart *shoppingCart = [self stringToShoppingCart:arg0];
    
    [TalkingDataAppCpa onViewShoppingCart:shoppingCart];
}

- (void)onPlaceOrder:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *arg1 = [command.arguments objectAtIndex:1];
    TDOrder *order = [self stringToOrder:arg1];
    
    [TalkingDataAppCpa onPlaceOrder:profileId withOrder:order];
}

- (void)onOrderPaySucc:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *orderId = [command.arguments objectAtIndex:1];
    if (![orderId isKindOfClass:[NSString class]]) {
        orderId = nil;
    }
    
    int amount = 0;
    NSNumber *arg2 = [command.arguments objectAtIndex:2];
    if ([arg2 isKindOfClass:[NSNumber class]]) {
        amount = [arg2 intValue];
    }
    
    NSString *currencyType = [command.arguments objectAtIndex:3];
    if (![currencyType isKindOfClass:[NSString class]]) {
        currencyType = nil;
    }
    
    NSString *payType = [command.arguments objectAtIndex:4];
    if (![payType isKindOfClass:[NSString class]]) {
        payType = nil;
    }
    
    [TalkingDataAppCpa onOrderPaySucc:profileId withOrderId:orderId withAmount:amount withCurrencyType:currencyType withPayType:payType];
}

- (void)onCredit:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    int amount = 0;
    NSNumber *arg1 = [command.arguments objectAtIndex:1];
    if ([arg1 isKindOfClass:[NSNumber class]]) {
        amount = [arg1 intValue];
    }
    
    NSString *content = [command.arguments objectAtIndex:2];
    if (![content isKindOfClass:[NSString class]]) {
        content = nil;
    }
    
    [TalkingDataAppCpa onCredit:profileId amount:amount content:content];
}

- (void)onTransaction:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *arg1 = [command.arguments objectAtIndex:1];
    TDTransaction *transaction = [self stringToTransaction:arg1];
    
    [TalkingDataAppCpa onTransaction:profileId transaction:transaction];
}

- (void)onCreateRole:(CDVInvokedUrlCommand*)command {
    NSString *roleName = [command.arguments objectAtIndex:0];
    if (![roleName isKindOfClass:[NSString class]]) {
        roleName = nil;
    }
    
    [TalkingDataAppCpa onCreateRole:roleName];
}

- (void)onLevelPass:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *levelId = [command.arguments objectAtIndex:1];
    if (![levelId isKindOfClass:[NSString class]]) {
        levelId = nil;
    }
    
    [TalkingDataAppCpa onLevelPass:profileId levelId:levelId];
}

- (void)onGuideFinished:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *content = [command.arguments objectAtIndex:1];
    if (![content isKindOfClass:[NSString class]]) {
        content = nil;
    }
    
    [TalkingDataAppCpa onGuideFinished:profileId content:content];
}

- (void)onLearn:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *course = [command.arguments objectAtIndex:1];
    if (![course isKindOfClass:[NSString class]]) {
        course = nil;
    }
    
    int64_t begin = 0;
    NSNumber *arg2 = [command.arguments objectAtIndex:2];
    if ([arg2 isKindOfClass:[NSNumber class]]) {
        begin = [arg2 longLongValue];
    }
    
    int duration = 0;
    NSNumber *arg3 = [command.arguments objectAtIndex:3];
    if ([arg3 isKindOfClass:[NSNumber class]]) {
        duration = [arg3 intValue];
    }

    [TalkingDataAppCpa onLearn:profileId course:course begin:begin duration:duration];
}

- (void)onPreviewFinished:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *content = [command.arguments objectAtIndex:1];
    if (![content isKindOfClass:[NSString class]]) {
        content = nil;
    }
    
    [TalkingDataAppCpa onPreviewFinished:profileId content:content];
}

- (void)onRead:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *book = [command.arguments objectAtIndex:1];
    if (![book isKindOfClass:[NSString class]]) {
        book = nil;
    }
    
    int64_t begin = 0;
    NSNumber *arg2 = [command.arguments objectAtIndex:2];
    if ([arg2 isKindOfClass:[NSNumber class]]) {
        begin = [arg2 longLongValue];
    }
    
    int duration = 0;
    NSNumber *arg3 = [command.arguments objectAtIndex:3];
    if ([arg3 isKindOfClass:[NSNumber class]]) {
        duration = [arg3 intValue];
    }
    
    [TalkingDataAppCpa onRead:profileId book:book begin:begin duration:duration];
}

- (void)onFreeFinished:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *content = [command.arguments objectAtIndex:1];
    if (![content isKindOfClass:[NSString class]]) {
        content = nil;
    }
    
    [TalkingDataAppCpa onFreeFinished:profileId content:content];
}

- (void)onAchievementUnlock:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *achievementId = [command.arguments objectAtIndex:1];
    if (![achievementId isKindOfClass:[NSString class]]) {
        achievementId = nil;
    }
    
    [TalkingDataAppCpa onAchievementUnlock:profileId achievementId:achievementId];
}

- (void)onBrowse:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *content = [command.arguments objectAtIndex:1];
    if (![content isKindOfClass:[NSString class]]) {
        content = nil;
    }
    
    int64_t begin = 0;
    NSNumber *arg2 = [command.arguments objectAtIndex:2];
    if ([arg2 isKindOfClass:[NSNumber class]]) {
        begin = [arg2 longLongValue];
    }
    
    int duration = 0;
    NSNumber *arg3 = [command.arguments objectAtIndex:3];
    if ([arg3 isKindOfClass:[NSNumber class]]) {
        duration = [arg3 intValue];
    }
    
    [TalkingDataAppCpa onBrowse:profileId content:content begin:begin duration:duration];
}

- (void)onTrialFinished:(CDVInvokedUrlCommand*)command {
    NSString *profileId = [command.arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    
    NSString *content = [command.arguments objectAtIndex:1];
    if (![content isKindOfClass:[NSString class]]) {
        content = nil;
    }
    
    [TalkingDataAppCpa onTrialFinished:profileId content:content];
}

- (void)onCustEvent1:(CDVInvokedUrlCommand*)command {
    [TalkingDataAppCpa onCustEvent1];
}

- (void)onCustEvent2:(CDVInvokedUrlCommand*)command {
    [TalkingDataAppCpa onCustEvent2];
}

- (void)onCustEvent3:(CDVInvokedUrlCommand*)command {
    [TalkingDataAppCpa onCustEvent3];
}

- (void)onCustEvent4:(CDVInvokedUrlCommand*)command {
    [TalkingDataAppCpa onCustEvent4];
}

- (void)onCustEvent5:(CDVInvokedUrlCommand*)command {
    [TalkingDataAppCpa onCustEvent5];
}

- (void)onCustEvent6:(CDVInvokedUrlCommand*)command {
    [TalkingDataAppCpa onCustEvent6];
}

- (void)onCustEvent7:(CDVInvokedUrlCommand*)command {
    [TalkingDataAppCpa onCustEvent7];
}

- (void)onCustEvent8:(CDVInvokedUrlCommand*)command {
    [TalkingDataAppCpa onCustEvent8];
}

- (void)onCustEvent9:(CDVInvokedUrlCommand*)command {
    [TalkingDataAppCpa onCustEvent9];
}

- (void)onCustEvent10:(CDVInvokedUrlCommand*)command {
    [TalkingDataAppCpa onCustEvent10];
}

- (NSDictionary *)jsonToDictionary:(NSString *)jsonStr {
    if (![jsonStr isKindOfClass:[NSString class]]) {
        return nil;
    }
    
    NSError *error = nil;
    id object = [NSJSONSerialization JSONObjectWithData:[jsonStr dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:&error];
    if (error || ![object isKindOfClass:[NSDictionary class]]) {
        return nil;
    }
    
    return object;
}

- (TDSearch *)stringToSearch:(NSString *)searchStr {
    NSDictionary *searchDic = [self jsonToDictionary:searchStr];
    if (searchDic == nil) {
        return nil;
    }
    
    TDSearch *search = [[TDSearch alloc] init];
    search.category = searchDic[@"category"];
    search.content = searchDic[@"content"];
    search.itemLocationId = searchDic[@"itemLocationId"];
    search.itemId = searchDic[@"itemId"];
    search.destination = searchDic[@"destination"];
    search.origin = searchDic[@"origin"];
    search.startDate = [searchDic[@"startDate"] intValue];
    search.endDate = [searchDic[@"endDate"] intValue];

    return search;
}

- (TDOrder *)stringToOrder:(NSString *)orderStr {
    NSDictionary *orderDic = [self jsonToDictionary:orderStr];
    if (orderDic == nil) {
        return nil;
    }
    
    TDOrder *order = [TDOrder orderWithOrderId:orderDic[@"orderId"] total:[orderDic[@"total"] intValue] currencyType:orderDic[@"currencyType"]];
    NSArray *items = orderDic[@"items"];
    for (NSDictionary *item in items) {
        [order addItemWithCategory:item[@"category"] itemId:item[@"itemId"] name:item[@"name"] unitPrice:[item[@"unitPrice"] intValue] amount:[item[@"amount"] intValue]];
    }
    
    return order;
}

- (TDShoppingCart *)stringToShoppingCart:(NSString *)shoppingCartStr {
    NSDictionary *shoppingCartDic = [self jsonToDictionary:shoppingCartStr];
    if (shoppingCartDic == nil) {
        return nil;
    }
    
    TDShoppingCart *shoppingCart = [TDShoppingCart createShoppingCart];
    NSArray *items = shoppingCartDic[@"items"];
    for (NSDictionary *item in items) {
        [shoppingCart addItemWithCategory:item[@"category"] itemId:item[@"itemId"] name:item[@"name"] unitPrice:[item[@"unitPrice"] intValue] amount:[item[@"amount"] intValue]];
    }
    
    return shoppingCart;
}

- (TDTransaction *)stringToTransaction:(NSString *)transactionStr {
    NSDictionary *transactionDic = [self jsonToDictionary:transactionStr];
    if (transactionDic == nil) {
        return nil;
    }
    
    TDTransaction *transaction = [[TDTransaction alloc] init];
    transaction.transactionId = transactionDic[@"transactionId"];
    transaction.category = transactionDic[@"category"];
    transaction.amount = [transactionDic[@"amount"]intValue];
    transaction.personA = transactionDic[@"personA"];
    transaction.personB = transactionDic[@"personB"];
    transaction.startDate = [transactionDic[@"startDate"] intValue];
    transaction.endDate = [transactionDic[@"endDate"] intValue];
    transaction.currencyType = transactionDic[@"currencyType"];
    transaction.content = transactionDic[@"content"];
    
    return transaction;
}

@end
